### PR TITLE
List the breaking changes from Jedis 3.x to Jedis 4.0

### PR DESCRIPTION
**Pull request by sazzad16**

### CODE REVIEW

**Review by dengliming — APPROVED** [review]

File: docs/3to4.md [hunk 1]
```diff
@@ -1,61 +1,61 @@
...
-- Jedis and JedisPool constructors with a String parameter without any int parameter will only support a URL or URI string.
+- Jedis and JedisPool constructors with a String parameter, and no int parameter only support a URL or URI string.
   - Jedis(String)
   - JedisPool(String)
   - JedisPool(String, SSLSocketFactory, SSLParameters, HostnameVerifier)
   - JedisPool(GenericObjectPoolConfig<Jedis>, String)
 
-- In xrange or xrevrange methods with StreamEntryID as parameters,
-  sending `null` for both start and end parameters would raise conflict.
+- [XRANGE](https://redis.io/commands/XRANGE) and [XREVRANGE](https://redis.io/commands/xrevrange) methods with StreamEntryID as parameters,
```
Comment (sazzad16) [inline docs/3to4.md:9]:
```suggestion
- In [XRANGE](https://redis.io/commands/XRANGE) and [XREVRANGE](https://redis.io/commands/xrevrange) methods with StreamEntryID as parameters,
```

File: docs/3to4.md [hunk 2]
```diff
@@ -1,61 +1,61 @@
...
+  sending `null` for both start and end parameters will raise conflicts.
```
Comment (sazzad16) [inline docs/3to4.md:10]:
```suggestion
  sending untyped `null` for both start and end parameters will raise conflicts.
```

File: docs/3to4.md [hunk 3]
```diff
@@ -97,44 +97,44 @@
...
+- Jedis.scriptExists(byte[]) returns `Boolean` instead of `Long`.
 
-- Jedis.scriptExists(byte[]...) now returns `List<Boolean>` instead of `List<Long>`.
+- Jedis.scriptExists(byte[]...) returns `List<Boolean>` instead of `List<Long>`.
 
-- `watch` and `unwatch` methods of Transaction class is not supported within MULTI (after multi() method). It can be used before MULTI.
+- `watch` and `unwatch` methods of Transaction class is unsupported within MULTI (after multi() method). It can be used before MULTI.
 
-- Following cases would throw IllegalStateException instead of JedisDataException.
+- The following cases would throw an IllegalStateException instead of a JedisDataException.
```
Comment (sazzad16) [inline docs/3to4.md:129]: Each of the following cases would throw an IllegalStateException instead of a JedisDataException.

File: docs/3to4.md [hunk 4]
```diff
@@ -1,61 +1,61 @@
...
-  Previously, in ideal scenario, it would always return null.
+- The return type of Jedis.shutdown() is now `void`.<br>
+  Previously, ideally, it would return null.
 
-- ShardedJedisPool, Sharded, ShardedJedis, BinaryShardedJedis, ShardInfo, JedisShardInfo classes are removed.
+- ShardedJedisPool, Sharded, ShardedJedis, BinaryShardedJedis, ShardInfo, JedisShardInfo classes have been removed.
   - Introduced JedisSharding class.
 
-      Earlier codes without the use of "name" and "weight" (in ShardInfo/JedisShardInfo) are easily transferable new class.
+      Earlier code without the use of "name" and "weight" (in ShardInfo/JedisShardInfo) are easily transferable new class.
```
Comment (sazzad16) [inline docs/3to4.md:32]: Earlier code without the use of "name" and "weight" (in ShardInfo/JedisShardInfo) are easily transferable to new class.

File: docs/3to4.md [hunk 5]
```diff
@@ -1,61 +1,61 @@
...
+- Commands interfaces were restructured, meaning `redis.clients.jedis.commands` package was reimplemented.
 
-- Return type of Jedis.shutdown() is now `void`.<br>
-  Previously, in ideal scenario, it would always return null.
+- The return type of Jedis.shutdown() is now `void`.<br>
+  Previously, ideally, it would return null.
 
-- ShardedJedisPool, Sharded, ShardedJedis, BinaryShardedJedis, ShardInfo, JedisShardInfo classes are removed.
+- ShardedJedisPool, Sharded, ShardedJedis, BinaryShardedJedis, ShardInfo, JedisShardInfo classes have been removed.
   - Introduced JedisSharding class.
```
Comment (sazzad16) [inline docs/3to4.md:30]: to replace ShardedJedisPool

File: docs/3to4.md [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+- Client and BinaryClient classes have been removed.
+
+- Commands interfaces were restructured, meaning `redis.clients.jedis.commands` package was reimplemented.
+
+- The return type of Jedis.shutdown() is now `void`.<br>
+  Previously, ideally, it would return null.
+
+- ShardedJedisPool, Sharded, ShardedJedis, BinaryShardedJedis, ShardInfo, JedisShardInfo classes have been removed.
+  - Introduced JedisSharding class.
```
Comment (sazzad16) [inline docs/3to4.md:30]:
```suggestion
  - Introduced JedisSharding class to replace ShardedJedisPool.
```

File: docs/3to4.md [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+      Earlier code without the use of "name" and "weight" (in ShardInfo/JedisShardInfo) are easily transferable new class.
```
Comment (sazzad16) [inline docs/3to4.md:32]:
```suggestion
      Earlier code without the use of "name" and "weight" (in ShardInfo/JedisShardInfo) are easily transferable to the new class.
```

File: docs/3to4.md [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+- ShardedJedisPipeline class was removed.
+  - Introduced ShardedPipeline class.
```
Comment (sazzad16) [inline docs/3to4.md:35]:
```suggestion
  - Introduced ShardedPipeline class to replace ShardedJedisPipeline.
```

File: docs/3to4.md [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+- Jedis.pubsubNumSub(String...) method returns `Map<String, Long>` instead of `Map<String, String>`.
+
+- Jedis.scriptExists(byte[]) returns `Boolean` instead of `Long`.
+
+- Jedis.scriptExists(byte[]...) returns `List<Boolean>` instead of `List<Long>`.
+
+- `watch` and `unwatch` methods of Transaction class is unsupported within MULTI (after multi() method). It can be used before MULTI.
+
+- The following cases would throw an IllegalStateException instead of a JedisDataException.
```
Comment (sazzad16) [inline docs/3to4.md:129]:
```suggestion
- Each of the following cases would throw an IllegalStateException instead of a JedisDataException.
```

**Comment by yangbodong22011** [comment]
@sazzad16 Jedis4 also add some new feature, such as `ClusterPipeline`, `RetryableCommandExecutor`(User can use to implement safe retry), do these need to be reflected in this document?

**Comment by sazzad16** [comment]
@yangbodong22011 It'd be great if we could add those and, if possible, we should. But those should be in new docs, not in the docs of this PR.

**Review by zeekling — DISMISSED** [review]

**Review by yangbodong22011 — APPROVED** [review]

**Review by dengliming — DISMISSED** [review]

**Review by yangbodong22011 — DISMISSED** [review]

**Review by chayim — APPROVED** [review]

**Pull request merged**